### ISSUE TITLE
Cache the promise resolving to the Offline Cache to increase performance

### ISSUE THREAD

**Issue by azasypkin** [gaia-components/serviceworkerware#22]
It looks like current implementation of  ```SimpleOfflineCache.ensureCache``` calls ```CacheStorage.open``` (via ```cacheHelper.getCache(this.cacheName)```)too often, for every ```fetch``` request. Shouldn't it be implemented like this instead?

```
SimpleOfflineCache.prototype.ensureCache = function soc_ensureCache() {
  if (!this.cachePromise) {
    this.cachePromise = cacheHelper.getCache(this.cacheName);
  }

  return this.cachePromise;
};
```

Currently I'm playing with ```serviceworkerware``` in Messages app & Nightly Desktop. We don't optimize scripts and styles there, so we have quite a lot of requests when page is opened, ~170 (or in real life we have lazy loader with ~35 scripts). I've tried to collect some measurements with ```performance.now``` to compare current implementation with proposed above:
```
SimpleOfflineCache.prototype.onFetch = function soc_onFetch(request, response) {
  .....
  var t0 = performance.now();
  return this.ensureCache().then(function(cache) {
    debug('Cache retrieval time: ' + (performance.now() - t0));
    .....
  });
};
```

And for current implementation I see a lot of numbers between ```120ms - 20ms```, for the proposed it's around ```0.03ms``` for every fetch request (hope I haven't messed up with my measurements, maybe you have better way to measure :)).

So what do you think @arcturus ? Is it safe/not-forbidden to store cache storage like I propose? 

I can provide a patch if you think it is worth it.

**Comment by arcturus** [comment]
Hi @azasypkin,
we were planning to remove the use of the `CacheHelper` and go directly using the `Cache` api. At the end of the day it wasn't adding much.

Actually we are removing it's usage in some parts of the code:
https://github.com/gaia-components/serviceworkerware/pull/16

I think we should get ride of the `CacheHelper` in the `SimpleOfflineCache` middleware. @lodr commented that he would like to go directly for asking always for opening the cache, but if the performance is impacted that negatively, definitey, caching the promise sounds like a good idea.

@lodr do you agree?

**Comment by lodr** [comment]
Nice catch @azasypkin ! Let's cache the cache.

Just kidding. I did not though opening the cache could be so expensive. Baku told me it's because it involves IPC. Totally agree.

**Comment by lodr** [comment]
Oh, anyway, I push for getting rid of CacheHelper anyway.

**Comment by azasypkin** [comment]
> Let's cache the cache.

Yeah, I know, it sounds like a non-sense :)

> Baku told me it's because it involves IPC. Totally agree.

My only concern is: is that possible that ```CacheStorage``` instance we cache in promise can be become kind of defunct/stale somehow so that we need to reopen it again? Or it's not the case at all @lodr?

**Comment by lodr** [comment]
Except for destroying it. I can not get other way for a cache to become invalid. So I think we can simply be careful when deleting caches.

**Comment by flaki** [comment]
> Just kidding. I did not though opening the cache could be so expensive. Baku told me it's because it involves IPC. Totally agree.

AFAIK this is because of e10s, the service worker runs in a different thread, while caches are implemented on the main thread, so accessing them requires the use of a [messaging protocol](https://blog.wanderview.com/blog/2014/12/08/implementing-the-serviceworker-cache-api-in-gecko/) between the two threads.

**Comment by arcturus** [comment]
This has been solved in #29
